### PR TITLE
feat(codeowners): POST, PUT, DELETE endpoints

### PR DESCRIPTION
**Pull request by NisanthanNanthakumar**
## Objective:
Before saving a record in the `sentry_projectcodeowners` table, we need to do some validation/transformations.
- we need to transform CODEOWNERS syntax into IssueOwner syntax. Then we can reuse existing logic to tranform IssueOwner syntax into the schema.
- we need to validate that all the usernames, team names and emails in the CODEOWNERS file has an association in Sentry.
- we need to transform the file paths in CODEOWNERS with the code path mappings in Sentry.
- we need to create a default record in `sentry_projectownership` table if none exists, to create default settings

## Test Plan:
I wrote unit tests for the grammar functions and endpoint integration tests

## Future:
We need to create a migration to remove the `organization_integration` column from `sentry_projectcodeowners` table. We already have a reference to it from `repository_project_path_config` foreignkey column and it is possible to update both tables independantly, increasing the chances of a drift in consistency.

### CODE REVIEW

File: src/sentry/api/endpoints/project_codeowners.py
Comment (scefali) [inline src/sentry/api/endpoints/project_codeowners.py:44]: @NisanthanNanthakumar what about the default value for `organization_integration_id`?

File: src/sentry/api/endpoints/project_codeowners.py
Comment (scefali) [inline src/sentry/api/endpoints/project_codeowners.py:54]: @NisanthanNanthakumar shouldn't we check the `UserEmail` table instead? That's generally we check for email matches in Senty.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (scefali) [inline src/sentry/api/endpoints/project_codeowners.py:62]: @NisanthanNanthakumar should we add a column on the `external_name`? Not urgent but if this table gets big, then it could geet expensive.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (scefali) [inline src/sentry/api/endpoints/project_codeowners.py:80]: @NisanthanNanthakumar Is this going to look funky if you have multiple error messages joined in on string? Maybe we can provide a list of errors here instead of a string? Not sure the best way of handling this. Alternatively, we could just send only one error message.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (scefali) [inline src/sentry/api/endpoints/project_codeowners.py:84]: @NisanthanNanthakumar another place we probably need to use the `UserEmail` table

**Review by scefali — COMMENTED** [review]
@NisanthanNanthakumar looks good overall, the only major thing I see is that we aren't using the `UserEmail` table when looking up emails. However, I did not really check the parsing logic for `parse_code_owners` so I hope the tests are solid here hahah.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:44]: @scefali 

> We need to create a migration to remove the organization_integration column from sentry_projectcodeowners table. We already have a reference to it from repository_project_path_config foreignkey column and it is possible to update both tables independantly, increasing the chances of a drift in consistency.

I'm going to make this `Future` PR now and merge it before this PR. It'll eliminate the need to handle `organization_integration_id`.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:62]: @scefali yeah we'll monitor with Sentry Performance and we can add an index if it becomes slow.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:80]: @scefali hmm, I think it's ok bc its the error for 1 field in the request body and it lists all the problems associated with that 1 field. 
I think sending 1 message at a time would be bad DX. I'd rather know all the problems with my request upfront, rather than piecemeal.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:84]: @scefali the email in `auth_user` is the primary email in `UserEmail`. I think its ok to use this here bc we know the user exists.

Comment (scefali) [reply src/sentry/api/endpoints/project_codeowners.py:84]: @NisanthanNanthakumar Good point, once we have the user we can use primary email here

File: src/sentry/api/endpoints/project_codeowners.py
Comment (scefali) [reply src/sentry/api/endpoints/project_codeowners.py:80]: @NisanthanNanthakumar ok but maybe we should add new line characters between the messages?

File: src/sentry/api/endpoints/project_codeowners.py
Comment (scefali) [inline src/sentry/api/endpoints/project_codeowners.py:79]: @NisanthanNanthakumar Nit: I would recommend just making an array of errors then calling `"\n".join` on the array at the end

**Review by scefali — APPROVED** [review]

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:79]: @scefali 👍 smart!

File: src/sentry/api/endpoints/project_codeowners.py
Comment (markstory) [reply src/sentry/api/endpoints/project_codeowners.py:90]: Should this organizationmember query have a condition for the current organization? This looks like it could cross organization boundaries.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (markstory) [reply src/sentry/api/endpoints/project_codeowners.py:64]: Should this have any conditions on the current organization or organization integration record?

File: src/sentry/api/endpoints/project_codeowners_details.py
Comment (markstory) [inline src/sentry/api/endpoints/project_codeowners_details.py:40]: Should this be here?

File: src/sentry/api/endpoints/project_codeowners_details.py
Comment (markstory) [inline src/sentry/api/endpoints/project_codeowners_details.py:27]: Should this query be scoped to the organization or projects that the current user is in? Right now I could delete mappings in another organization by enumerating the id values in the URL, but providing my project & organization.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:64]: @markstory yes ExternalUser has a `organizationmember_id` column and I'll add an inner join to the `organizationmember` table on `organization_id`.

File: src/sentry/api/endpoints/project_codeowners_details.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners_details.py:27]: @markstory ohhh I didn't think of that! I'll update the query for this.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:96]: @markstory I updated the query for `external_teams` to check for the current organization here: https://github.com/getsentry/sentry/blob/7df73c982e1f8660ed680bd5b4e41ca18ca5063e/src/sentry/api/endpoints/project_codeowners.py#L64-L67

So in users_dict, all the users are in the organization of the project.

File: src/sentry/api/endpoints/project_codeowners_details.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners_details.py:40]: @markstory I think it should because it is a path param for this route.
https://github.com/getsentry/sentry/blob/7df73c982e1f8660ed680bd5b4e41ca18ca5063e/src/sentry/api/urls.py#L1678-L1682

Comment (markstory) [reply src/sentry/api/endpoints/project_codeowners_details.py:40]: I meant the referenced to 'file'

Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners_details.py:40]: ahh I'll replace file with `CODEOWNERS`

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:126]: @markstory at this point in the logic flow, the ownership comes from here: https://github.com/getsentry/sentry/blob/7df73c982e1f8660ed680bd5b4e41ca18ca5063e/src/sentry/api/endpoints/project_ownership.py#L103-L107

And the values cannot be null. I don't want to modify this mixin, bc its used elsewhere. Do you think I should overwrite the `get_ownership` method and set the default time there?

File: src/sentry/api/endpoints/project_codeowners_details.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners_details.py]: Updated and added a test to confirm the response.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (markstory) [reply src/sentry/api/endpoints/project_codeowners.py:126]: Oh. I don't know where else that method is used and changing the mixin could have other side effects.

Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:126]: @markstory yeah then for now, I think its ok to keep as is and set the time if its null

**Comment by NisanthanNanthakumar** [comment]
This PR was put on hold until https://github.com/getsentry/sentry/pull/24163 got merged. Now this is ready.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (MeredithAnya) [inline src/sentry/api/endpoints/project_codeowners.py:58]: I wonder if we could pull out some of the logic into like `_validate_association()`

```python
def _validate_association(raw_items, associations, type):
    if type == "emails":
         sentry_items = [item.email for item in associations]
    else: 
         sentry_items = [item.external_name for item in associations]

    diff = [item for item in raw_items if not item not in sentry_items]

    if len(diff):
        return [f'The following %s do not have an association in Sentry: {", ".join(diff)}]
    
    return []
```
    
where the raw items would be the values from the `parse_code_owners` the types would be `emails`, `usernames` `teamnames` and associations would be the sentry querysets

we don't have to do something like this but it does feel like there is a lot of repeated code that is very similar. Also I would say we could raise early, but im guessing that we rather give them all the missing associations instead of just the first set so that they can fix them all at once

File: src/sentry/api/endpoints/project_codeowners.py
Comment (MeredithAnya) [inline src/sentry/api/endpoints/project_codeowners.py:131]: Do we expect the `code_mapping_id` not to be there? The `ProjectCodeOwners` isn't going to save without it so should we be raising an error?

File: src/sentry/api/endpoints/project_codeowners_details.py
Comment (MeredithAnya) [inline src/sentry/api/endpoints/project_codeowners_details.py:29]: Do we normally return the id of the object they are trying to find?

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:30]: @MeredithAnya It looks like it's the default only when `default=empty`. But there is precedence with other serializers being explicit about `required=True`. I think it is better to be explicit for future maintenance and if there are changes to the `default` param.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:58]: @MeredithAnya Good point on the duplicated filter logic! Yeah I don't think it'll be a good DX to raise early. I think its preferable to return all the missing associations at once and let them fix at once.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:131]: @MeredithAnya hmm, I don't expect "code_mapping_id" to never be there. It would have been caught earlier in the validation flow. Yes, also Postgres will throw a constraint error if we try to save without it. 

The None is just an explicit default to prevent a KeyError. Is this not pythonic?

File: src/sentry/api/endpoints/project_codeowners_details.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners_details.py:29]: Hmmm, I guess not. Is there any reason we shouldn't? I think a descriptive error message saying exactly whats wrong with the request is better developer experience. I don't think the id is privileged info, since its part of the request url.

Comment (MeredithAnya) [reply src/sentry/api/endpoints/project_codeowners_details.py:29]: Well it could also be the wrong `id` _with_ that particular `project`. I think it's more accurate to just say like we couldn't find it imo

File: src/sentry/api/endpoints/project_codeowners.py
Comment (MeredithAnya) [reply src/sentry/api/endpoints/project_codeowners.py:139]: I guess my question is, why prevent a `KeyError` there just to get an `IntegrityError` later?

File: src/sentry/api/endpoints/project_codeowners_details.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners_details.py]: @MeredithAnya 👍  I also found that we are mostly using the `ResourceDoesNotExist` exception for this case. So I updated  ProjectCodeOwners to stick with our current convention.

File: src/sentry/api/endpoints/project_codeowners.py
Comment (MeredithAnya) [inline src/sentry/api/endpoints/project_codeowners.py:79]: I don't think you have to check the `len` now cause `external_association_err.extend([])` will still keep it an empty array

**Review by MeredithAnya — APPROVED** [review]
Left one more comment that's more a nit but I think this looks good

File: src/sentry/api/endpoints/project_codeowners.py
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_codeowners.py:139]: hmm good question! To preface this discussion, we will never get to this point bc if `code_mapping_id` is missing, it would be caught by the validation flow.

`KeyError` is application-level and `IntegrityError` is db-level. I think there is more detail/context provided by an IntegrityError than KeyError. KeyError says that the key does not exist in the dictionary and thats it. But IntegrityError would provide us with  details on which column it failed to save on and why.

**Pull request merged**
